### DISCUSSION TITLE
The problem of mesh refinement

### DISCUSSION THREAD

**Discussion by PEI0214** [idaholab/moose#29079] [Q&A General]
### Check these boxes if you have followed the posting rules.

- [X] Q&A General is the most appropriate section for my question
- [X] I have consulted the posting Guidelines on the Discussions front page
- [X] I have searched the Discussions forum and my question has not been asked before
- [X] I have searched the MOOSE website and the documentation does not answer my question
- [X] I have formatted my post following the posting guidelines (screenshots as a last resort, triple back quotes around pasted text)

### Question

Hello,
When I use mesh refinement and MeshModifiers together, the following happens. This doesn't seem normal. When the grid is 8-node cells, this phenomenon does not occur.
<details>

<summary>Grid and result</summary>

The 6-node grid and results are as follows:
![image](https://github.com/user-attachments/assets/26e025bb-1c13-44ea-85b9-63f82304b67e)
![image](https://github.com/user-attachments/assets/3b808b26-f739-46b9-a97a-327921d109c2)

The 8-node grid and results are as follows:
![image](https://github.com/user-attachments/assets/fb8c01e1-02f6-4453-89b8-4454eda050ab)
![image](https://github.com/user-attachments/assets/de9470c8-f1ff-4cd5-bc63-0341c3cb24c4)

</details>
The setting of boundary conditions and mesh refinement are the same. It's all full restraint on the bottom and pressure on the top. Why does this happen?

**Comment by GiudGiud** [comment]
Hello

This looks like the solve is failing and the final value for mesh displacements in unphysical. 
Can you run these other cases:
- no mesh modifier, no mesh refinement
- mesh modifier, no mesh refinement
- no mesh modifier, mesh refinement
To see if the interaction of the two is the problem.

Also can you please paste your whole input

**Reply by PEI0214** [reply]
There is no problem with either one alone, only when used together. Here is my input file:
[mesh.zip](https://github.com/user-attachments/files/17762262/mesh.zip)

<details>

<summary>input</summary>

```
[Problem]
  kernel_coverage_check = false
  material_coverage_check = false
  boundary_restricted_node_integrity_check = false
  boundary_restricted_elem_integrity_check = false
[]

new_domain_blocks = '1001    1002'

[Mesh]
  [box]
    type = FileMeshGenerator
    file = mesh4.msh
  []

  [refine]
    type = RefineBlockGenerator
    input = box
    block = '2'
    refinement = '1'
    enable_neighbor_refinement = false
  []

  add_subdomain_ids = '${new_domain_blocks}'
[]

[GlobalParams]
  displacements = 'disp_x disp_y disp_z'
[]

[Variables]
  [disp_x]
    block = '${new_domain_blocks}'
  []
  [disp_y]
    block = '${new_domain_blocks}'
  []
  [disp_z]
    block = '${new_domain_blocks}'
  []
[]

# ===== Tensor mechanics for all active domain blocks =====
[Physics]
  [SolidMechanics]
    [QuasiStatic]
      [all]
        strain = finite
        add_variables = false
        block = '${new_domain_blocks}'
        generate_output = 'stress_xx stress_zz'
      []
    []
  []
[]

# ===== Gravity =====
[Kernels]
  [Gravity1]
    type = Gravity
    block = '${new_domain_blocks}'
    use_displaced_mesh = false
    variable = disp_z
    value = -10
  []
[]

# ===== Boundary Conditions =====
[BCs]
  [archor_x]
    type = DirichletBC
    boundary = '3'
    variable = disp_x
    value = 0
  []

  [archor_y]
    type = DirichletBC
    boundary = '3'
    variable = disp_y
    value = 0
  []

  [archor_z]
    type = DirichletBC
    boundary = '3'
    variable = disp_z
    value = 0
  []

  [dz]
    type = FunctionNeumannBC
    boundary = '4'
    variable = disp_z
    function = '-2e3 * (t - 200)'
  []
[]

# ===== Materials (linear-elastic to keep it simple) =====
[Materials]
  [elasticity_tensor1]
    type = ComputeIsotropicElasticityTensor
    block = '${new_domain_blocks}'
    youngs_modulus = 50E6 # 50 MPa
    poissons_ratio = 0.3
  []

  [stress]
    type = ComputeFiniteStrainElasticStress
    block = '${new_domain_blocks}'
  []

  [density1]
    type = GenericConstantMaterial
    block = '${new_domain_blocks}'
    prop_names = density
    prop_values = 2000
  []
[]

[MeshModifiers]
  [m1]
    type = TimedSubdomainModifier
    times = '100 200'
    blocks_from = '1 2'
    blocks_to = '1001    1002'
    execute_on = 'TIMESTEP_BEGIN'
  []
[]

[Controls]
  [c1]
    type = TimePeriod
    enable_objects = 'BCs::dz'
    start_time = '200'
  []
[]

# ===== Executioner =====
[Executioner]
  type = Transient

  end_time = 500
  dt = 100

  solve_type = 'PJFNK'
  petsc_options = '-snes_converged_reason'

  petsc_options_iname = '-ksp_gmres_restart -pc_type -pc_hypre_type -pc_hypre_boomeramg_max_iter'
  petsc_options_value = '  201               hypre    boomeramg      10'

  nl_abs_tol = 1E-5
  nl_max_its = 400

  l_tol = 1E-8
  l_max_its = 200
[]

[Outputs]
  exodus = true
[]
```

</details>
Is there any literature on mesh adaptation in MOOSE?

**Reply by GiudGiud** [reply]
There s the adaptivity syntax page that gives good information about it.

**Reply by GiudGiud** [reply]
I ll take a look at your inputs tomorrow. Can you please put the mesh file on a public GitHub repo? Zip files are iffy to download

**Reply by PEI0214** [reply]
The grid is of the gmsh type.
<details>

<summary>mesh4.msh</summary>

```
$MeshFormat
2.2 0 8
$EndMeshFormat
$PhysicalNames
4
3	1	vol_1
3	2	vol_2
2 3 "bottom"
2 4 "top"
$EndPhysicalNames
$Nodes
9
1 0 0 0
2 1 0 0
3 0 1 0
4 0 0 0.2
5 1 0 0.2
6 0 1 0.2
7 0 0 0.4
8 1 0 0.4
9 0 1 0.4
$EndNodes
$Elements
4
1 6 2 1 11 1 2 3 4 5 6
2 6 2 2 12 4 5 6 7 8 9
3 2 2 3 13 1 2 3
4 2 2 4 14 7 8 9
$EndElements

```

</details>

**Reply by PEI0214** [reply]
@GiudGiud 
Hello. Have you tried to run it yet?

**Reply by GiudGiud** [reply]
not yet!

**Reply by PEI0214** [reply]
Hello, I have one more question. After MOOSE uses adaptive generation of non-uniform mesh, how does it compute? How are the nodes in the figure below handled? I couldn't find the description on MOOSE's website. Is there any literature to answer this question?
![image](https://github.com/user-attachments/assets/866d6851-9083-4a3c-9b44-4987f679955a)

**Reply by GiudGiud** [reply]
This is more of a libmesh question actually. The short answer is that the dofs on those nodes are constrained to ensure continuity (for continuous variables)

**Reply by PEI0214** [reply]
Is there more information to learn about it? Can you run the input file above? It has problems in the calculation of refined mesh.

**Reply by PEI0214** [reply]
Use MeshModifiers for six-node units after refinement.  The constraint on the hanging node seems to have failed.

**Reply by GiudGiud** [reply]
?
how so?

**Reply by PEI0214** [reply]
This will get the wrong result.
![image](https://github.com/user-attachments/assets/d35c162f-30ab-4a0d-8672-85188fc790aa)

**Reply by GiudGiud** [reply]
Can you try setting the boundary conditions to use preset= false

And is the nodeset properly covering the newly appeared nodes from refinement ? What does the output show

**Comment by PEI0214** [comment]
>Can you try setting the boundary conditions to use preset= false

I tried setting preset = false, but it didn't affect the result.

>And is the nodeset properly covering the newly appeared nodes from refinement ? What does the output show

How to check the nodeset? It should be correct.
![f00d628dac364a35f9abb15a00e8bc7b](https://github.com/user-attachments/assets/40905259-719a-4fe8-9663-f41cc99225d6)

**Reply by GiudGiud** [reply]
in paraview, on the left pane you should be able to show/hide blocks, and if you click the advanced settings you can do the same for nodes
I would like to be sure that the top nodeset contains 8 7 0 and 19 20 21 and no other

**Reply by PEI0214** [reply]
This is the top nodeset.
![530a4f9376086b3ca7db66e57418762a](https://github.com/user-attachments/assets/3c7c3202-68a6-478f-bbe4-f7249a06b208)

**Reply by GiudGiud** [reply]
ok this looks right. 
I ll take a look when I have time. I am at a conference right now

**Reply by PEI0214** [reply]
ok, thank you very much!

**Reply by GiudGiud** [reply]
So first
```
    enable_neighbor_refinement = false
```
is a bad idea. As mentioned in the documentation, neighbor refinement is there to avoid problems. 
Let's keep it to the default


Looks like the block restriction of the displacements is the problem. Using this makes it work for me
```

[Variables]
  [disp_x]
    # block = '${new_domain_blocks}'
  []
  [disp_y]
    # block = '${new_domain_blocks}'
  []
  [disp_z]
    # block = '${new_domain_blocks}'
  []
[]

# Present empty equations before subdomains are changed
[Kernels]
  [n1]
    type = NullKernel
    block = '1 2'
    variable = disp_x
  []
  [n2]
    type = NullKernel
    block = '1 2'
    variable = disp_y
  []
  [n3]
    type = NullKernel
    block = '1 2'
    variable = disp_z
  []
```

**Reply by PEI0214** [reply]
In my simulation, the mesh needs to be gradually activated. If I don't set block restriction of the displacements, the inactive mesh will also have displacements, which leads to poor results. Is there a way to avoid this? While using displacement block restriction, the mesh should still be refined properly.

**Reply by GiudGiud** [reply]
>  the inactive mesh will also have displacements, which leads to poor results.

It should not, not with just the Nullkernels as the equation for displacements

**Reply by PEI0214** [reply]
Does this require creating a kernel for zero displacement?

**Reply by GiudGiud** [reply]
I dont think so. See the input snippet I posted.
But you can try that if you think that will help. Having kernels that push the displacements to 0 in a region where they should be 0 will not hurt
